scipy/special/_basic.py

```diff
@@ -2372,51 +2447,54 @@ def factorial(n, exact=False):
     120
 
     """
-    if exact:
-        if np.ndim(n) == 0:
-            if np.isnan(n):
-                return n
-            return 0 if n < 0 else math.factorial(n)
-        else:
-            n = asarray(n)
-            un = np.unique(n).astype(object)
-
-            # Convert to object array of long ints if np.int_ can't handle size
-            if np.isnan(n).any():
-                dt = float
-            elif un[-1] > 20:
-                dt = object
-            elif un[-1] > 12:
-                dt = np.int64
+    # handle array-likes
+    n = n if np.isscalar(n) else asarray(n)
+    if isinstance(n, np.ndarray):
+        if not (np.issubdtype(n.dtype, np.integer)
+                or np.issubdtype(n.dtype, np.floating)):
+            raise ValueError(f"Unsupported datatype for factorial: {n.dtype}")
```

Comment:
The error should also say what is supported to be helpful.

Reply:
Good point